archive/imo/imo1969_q1.lean

```diff
@@ -56,15 +55,30 @@ infinitely many of them.
 $a = 4*(2+b)^4$ should do. So $m = 2+b$.
 -/
 
-theorem imo1969_q1 : ∀ b : ℕ, ∃ a : ℕ, a ≥ b ∧ ∀ n : ℕ, ¬ prime (n^4 + a) :=
-assume b,
-have h1 : 1 < 2+b, by linarith,
-have b^2 ≥ b, by nlinarith,
-have h2 : 4*(2+b)^4 ≥ b, by nlinarith,
+/-- `good_nats` is an abbreviation for the subtype of natural numbers satisfying the
+condition in the problem statement, namely the `a : ℕ` such that `n^4 + a` is not prime
+for any `n : ℕ`. -/
+abbreviation good_nats : Type := {a : ℕ // ∀ n : ℕ, ¬ prime (n^4 + a)}
```

Comment:
I think that a set and `set.infinite` better reflects the original problem.

Reply:
A good lemma for `src/` (`not_congr finite_image_iff`): if `f : α → β` is `inj_on s`, then `f '' s` is an infinite set iff `s` is an infinite set. Related lemmas:

* if `f` is `inj_on s` and `s` is infinite and `maps_to f s t`, then `t` is infinite (above + monotonicity);
* if `f` is injective and `[infinite α]`, then `range f` is an infinite set;
* if `f` is injective, `[infinite α]` and `∀ x, f x ∈ s`, then `s` is an infinite set (this is what's I'd use in this problem).

Reply:
Done, see the additions to `data.set.finite`.